src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/AbstractAddItemCommand.cs

```diff
@@ -63,25 +67,18 @@ protected override Task<CommandStatusResult> GetCommandStatusAsync(IProjectTree
         {
             var nodeToAddTo = GetNodeToAddTo(node);
 
-            // Publish any existing changes that could potentially be here.
-            // Then call OnAddingNodesAsync.
+            // Call OnAddingNodesAsync.
             // Then publish changes that could have taken place in OnAddingNodesAsync.
-            await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
             await OnAddingNodesAsync(nodeToAddTo).ConfigureAwait(true);
             await _projectTree.TreeService.PublishLatestTreeAsync(waitForFileSystemUpdates: true).ConfigureAwait(true);
 
-            // Get the difference to see what was added.
-            // We do a sanity check to make sure they have a valid display order.
-            // We also order the added nodes by their display order.
             var updatedNode = _projectTree.CurrentTree.Find(node.Identity);
             var updatedNodeToAddTo = _projectTree.CurrentTree.Find(nodeToAddTo.Identity);
-            var addedNodes =
-                updatedNodeToAddTo.Children.Where(x => !nodeToAddTo.TryFind(x.Identity, out var subtree) && OrderingHelper.HasValidDisplayOrder(x))
-                .OrderBy(OrderingHelper.GetDisplayOrder).ToList();
+            var addedNodes = OrderingHelper.GetAddedNodes(nodeToAddTo, updatedNodeToAddTo);
 
             if (addedNodes.Any())
             {
-                await OnAddedNodesAsync(_projectVsServices.ActiveConfiguredProject, addedNodes, updatedNode).ConfigureAwait(true);
+                await OnAddedNodesAsync(_projectVsServices.ActiveConfiguredProject, node, addedNodes, updatedNode).ConfigureAwait(true);
```

Comment:
Why are we switching back to the original threads in this method? It's very unclear - and should have a comment. Normally we only switch back if 1) we need to stay on the UI thread, or 2) we took a lock and need to get back on the same thread.

Reply:
There isn't a reason why we are switching back. I'm happy to change it to "false".

Reply:
Make note, you'll need "true" if you are within the callback that you call within the write lock, but it doesn't appear you are.

Reply:
Yes. I think the reasoning behind setting it to "true" before was to accidentally prevent bad things from ever happening.